Resolves #281: Add some convenience APIs to FDBMetaDataStore

amouehsan opened the pull request:
This change adds addIndex, dropIndex and updateRecords to FDBMetaDataStore.
Implementing addIndex and dropIndex is fairly straightforward. For updateRecords,
we decided to add the following two functionalities to RecordMetaDataBuilder:
- Add new record types and process extensions for them if user asks for it.
- Update the descriptors of the existing unionFields and record types.

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStoreTest.java:
```diff
@@ -276,4 +279,69 @@ public void withToProto() throws Exception {
+            assertNotNull(metaDataStore.getRecordMetaData().getRecordType("MyOtherRecord"));
+            assertNotNull(metaDataStore.getRecordMetaData().getRecordType("AnotherRecord"));
+            context.commit();
+        }
```
Might be worth creating a new meta-data store in a new transaction so that it doesn't just read the meta-data from cache. (It looks like the `indexes` tests are already doing that.)

amouehsan replied:
good idea. sure.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+            RecordMetaDataBuilder recordMetaDataBuilder = createMetaDataBuilder(metaDataProto);
+            recordMetaDataBuilder.addIndex(recordType, index);
+            recordMetaData = recordMetaDataBuilder.build(true);
+            saveWithSplitHelper(recordMetaData);
```
I can do it. I didn't call `saveAndSetCurrent` or `saveRecordMetaData` because it unnecessarily loads. If it's cached, probably it's fine. but when it's not there yet, we would unnecessarily do a look up. Maybe I can keep this, and then later change `saveAndSetCurrent` or `saveRecordMetaData` to call this too? What do you think?

Fixed the `maintainHistory` issue.

amouehsan replied:
Actually thinking thru this again. In order to properly keep history, probably it would better to call saveAndSetCurrent. I'll apply your comment. thanks.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+            saveWithSplitHelper(recordMetaData);
+            return null;
+        });
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADD_INDEX, future);
```
sure.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+    }
+
+    /**
+     * Remove the index from the record meta-data.
```
done.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+
+    /**
+     * Update the meta-data records descriptor.
+     * As a result, it adds new record types and replaces all of the current descriptors with the new descriptors.
```
done

amouehsan commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java:
```diff
@@ -306,4 +306,14 @@ public void noPrimaryKey() {
+        RecordMetaData recordMetaData = builder.build(true);
+        assertNotNull(recordMetaData.getRecordType("MySimpleRecord"));
+        assertNotNull(recordMetaData.getRecordType("MyOtherRecord"));
+        assertNotNull(recordMetaData.getRecordType("AnotherRecord"));
```
sure.

alecgrieser replied:
It looks like there still aren't assertions about the `since_version`.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+        addIndex(recordType, new Index(indexName, indexExpression));
+    }
+
+    private void addIndex(@Nonnull String recordType, @Nonnull Index index) {
```
didn't see this. sure.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -428,6 +436,37 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
+    public void updateRecords(Descriptors.FileDescriptor recordsDescriptor, boolean processExtensionOptions) {
+        this.recordsDescriptor = recordsDescriptor;
+        unionDescriptor = updateRecordTypes(this.recordsDescriptor, processExtensionOptions);
+        processSchemaOptions(processExtensionOptions);
```
Yeah. The reason not to process the schema options is that there are already ways of setting them in the builder (i.e., `setSplitLongRecords` and `setStoreRecordVersions`), and I think we want to reserve this method for doing things that *cannot* be done through other mechanisms (which is to say, updating the actual message descriptors of the various record types) to avoid having the user accidentally mutate other things.

I think the one exception to that rule would be adding indexes and primary key information from the new record types based on those types extension information...but even then, I'm somewhat lukewarm to that idea.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -642,7 +681,28 @@ private static boolean unionHasMessageType(@Nonnull Descriptors.Descriptor union
+                }
+                // New field / record type.
+                processRecordType(unionField, processExtensionOptions);
+                unionFields.put(descriptor, unionField);
```
> > Doesn't this need to set the since version of the record type it creates
> 
> `processRecordType` does it, right?

If I'm reading `processRecordType` correctly, it only does so if the `since_version` is included in an extension option of the record type, but that extension option should be considered deprecated. (Hm, maybe it's worth validating that the since version in the option is the same as the since version we're setting if it's defined?) In fact, I think this is effectively supposed to replace setting the `since_version` of a record type as the easiest way for users to make sure that that field is set properly.

> > (and probably bump the meta-data version)
> 
> Good point. `RecordMetaDataBuilder` never bumped the meta-data version when a new record type is added. But may be we should for `updateRecords`. Bumping it is definitely safe, I guess I don't know if there's any problem with **not** bumping it though. If the record types have no indexes, what would bumping it give us?

If the record type doesn't have any indexes, bumping the version doesn't give us much (and neither does setting the `since_version`, I suppose). But after updating the records, the user very well might add some new indexes on the new record type. (We can't know that here either way.) And if they do, making sure the `since_version` is set is necessary to to avoid triggering unnecessary index builds on the indexes on the new types.

In theory, we only need to bump it once for all records. (It's fine if all of the records have the same `since_version`.) But for the `since_version` to be valid, it should be strictly greater than the meta-data version prior to the meta-data  version before the records were updated. (Otherwise, it will claim that it was added to the meta-data prior to the version in which it was actually added.)

It's also good practice to bump the version whenever the meta-data is changed so that those changes can be updated, so maybe bumping it once in `updateRecords` is sufficient.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -242,13 +247,17 @@ public void setMaintainHistory(boolean maintainHistory) {
+    protected RecordMetaDataBuilder createMetaDataBuilder(RecordMetaDataProto.MetaData metaDataProto) {
         RecordMetaDataBuilder builder = RecordMetaData.newBuilder().addDependencies(dependencies);
         if (localFileDescriptor != null) {
             builder.setLocalFileDescriptor(localFileDescriptor);
```
I the user has initialized the `FDBMetaDataStore` with a `localFileDescriptor`, won't this mean that the `updateRecords` method fails (somewhat mysteriously)? I think we would want it to create the meta-data without the local file descriptor for the `updateRecords` call and then (maybe) add it back in in `saveAndSetCurrent`.

amouehsan replied:
I actually think we should fail `updateRecords` if `localFileDescriptor` is set. I.e., `updateRecords` is not supported when `localFileDescriptor` is used and there shouldn't be any exception here either. Maybe that means we should not support `updateRecords` in `FDBMetaDataStore` at all. What do you think?

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +429,181 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
         context.asyncToSync(FDBStoreTimer.Waits.WAIT_SAVE_META_DATA, saveAndSetCurrent(metaDataProto));
     }
 
+    private CompletableFuture<RecordMetaDataProto.MetaData> loadCurrentProto() {
```
Missing nullability annotation

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
         /** Wait for an {@link IndexOperation} to complete. */
         WAIT_INDEX_OPERATION("wait for index operation"),
+        /** Wait for adding an index. */
+        WAIT_ADD_INDEX("wait for adding index to record meta-data"),
```
```suggestion
        WAIT_ADD_INDEX("wait for adding an index to a record meta-data"),
```

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
+        /** Wait for adding an index. */
+        WAIT_ADD_INDEX("wait for adding index to record meta-data"),
+        /** Wait for dropping an index. */
+        WAIT_DROP_INDEX("wait for dropping index from record meta-data"),
```
```suggestion
        WAIT_DROP_INDEX("wait for dropping an index from a record meta-data"),
```

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
+        /** Wait for dropping an index. */
+        WAIT_DROP_INDEX("wait for dropping index from record meta-data"),
+        /** Wait for updating records descriptor. */
+        WAIT_UPDATE_RECORDS_DESCRIPTOR("wait for updating the records descritpr in a record meta-data"),
```
```suggestion
        WAIT_UPDATE_RECORDS_DESCRIPTOR("wait for updating the records descriptor in a record meta-data"),
```

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java:
```diff
@@ -306,4 +306,21 @@ public void noPrimaryKey() {
+    public void updateRecords() {
+        RecordMetaDataBuilder builder = RecordMetaData.newBuilder().setRecords(TestRecords1Proto.getDescriptor());
+        assertNotNull(builder.getRecordType("MySimpleRecord"));
+        assertTrue(builder.getRecordType("MySimpleRecord").getDescriptor().getFile() == TestRecords1Proto.getDescriptor());
```
`assertSame(a, b)` instead of `assertTrue(a == b)` here and elsewhere. They amount to the same thing, but the error messages are much more understandable using `assertSame`.

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java:
```diff
@@ -306,4 +306,21 @@ public void noPrimaryKey() {
+        assertTrue(builder.getRecordType("MyOtherRecord").getDescriptor().getFile() == TestRecords1EvolvedProto.getDescriptor());
+        assertNotNull(recordMetaData.getRecordType("AnotherRecord"));
+        assertTrue(builder.getRecordType("AnotherRecord").getDescriptor().getFile() == TestRecords1EvolvedProto.getDescriptor());
+    }
```
Probably worth validating that calling `updateRecords` when `local_file_descriptor` is already set results in an error.

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStoreTest.java:
```diff
@@ -276,4 +280,135 @@ public void withToProto() throws Exception {
+    }
+
+    @Test
+    public void updateRecords() {
```
I think a test that combines `updateRecords` with a local file descriptor would be nice.

alecgrieser replied:
Where?

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +429,181 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+    }
+
+    /**
+     * Adds a new index to the record meta-data that contains multiple record types.
```
Roughly half of the methods here say "Adds a new index" and the other half say "Add a new index". I don't know which one I prefer, but I think they should all be the same.

amouehsan commented:
I preferred to keep the revision commits for the ease of review, but it was hard to resolve the merge conflicts. Squashed the revisions, rebased against master, and resolved the conflicts.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -435,6 +444,53 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
+     * <p>
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
+     * By contract, the extension options will be processed for all of the new record types and will not be processed
+     * for any of the old record types. Also, it is not allowed to call this method when the local file descriptor is set.
```
Given that the other overload of this function now has more information in its Javadoc description, I feel like this should link to the other one.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -435,6 +444,53 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
+     * Update the records descriptor of the record meta-data.
+     *
+     * <p>
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
```
```suggestion
     * This adds any new record types and updates the descriptors for existing record types.
```

I think we want to keep the union fields updating as an internal implementation detail. I could be wrong though.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -435,6 +444,53 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
+     * <p>
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
+     * By contract, the extension options will not be processed for the old record types. If {@code processExtensionOptions}
+     * is set, extension options will be processed only for the new record types. Also, it is not allowed to call this
```
```suggestion
     * is set, extension options will be processed only for the new record types. Also, this method may not be called
```

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java:
```diff
@@ -435,6 +444,53 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
+     * By contract, the extension options will not be processed for the old record types. If {@code processExtensionOptions}
+     * is set, extension options will be processed only for the new record types. Also, it is not allowed to call this
+     * method when the local file descriptor is set. This method bumps the meta-data version and sets {@code since_version}
```
```suggestion
     * when the local file descriptor is set. This method bumps the meta-data version and sets the {@linkplain RecordType#getSinceVersion() since version}
```

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java:
```diff
@@ -339,4 +340,33 @@ public void noPrimaryKey() {
+        assertSame(builder.getRecordType("MyOtherRecord").getDescriptor().getFile(), TestRecords1EvolvedProto.getDescriptor());
+        assertNotNull(recordMetaData.getRecordType("AnotherRecord"));
+        assertSame(builder.getRecordType("AnotherRecord").getDescriptor().getFile(), TestRecords1EvolvedProto.getDescriptor());
+        assertSame(builder.getRecordType("AnotherRecord").getSinceVersion(), builder.getVersion());
```
`assertEquals` rather than `assertSame` here, I think (for the latter assertion).

alecgrieser replied:
I think that's a reason to *not* to use `assertSame`. That assertion verifies that things are reference-equal rather than logically equal, and I think we only care about logical equality for this value.

If you cast `getSinceVersion` to an `int`, I think you can use `assertEquals`.

amouehsan replied:
sure, done.

alecgrieser commented on fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java:
```diff
@@ -339,4 +340,33 @@ public void noPrimaryKey() {
+        assertNotNull(recordMetaData.getRecordType("AnotherRecord"));
+        assertSame(builder.getRecordType("AnotherRecord").getDescriptor().getFile(), TestRecords1EvolvedProto.getDescriptor());
+        assertSame(builder.getRecordType("AnotherRecord").getSinceVersion(), builder.getVersion());
+        assertTrue(builder.getRecordType("AnotherRecord").getSinceVersion() > prevVersion);
```
```suggestion
        assertThat(builder.getRecordType("AnotherRecord").getSinceVersion(), greaterThan(prevVersion));
```

Accepting this suggestion will involve importing the right matcher.

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
         /** Wait for an {@link IndexOperation} to complete. */
         WAIT_INDEX_OPERATION("wait for index operation"),
+        /** Wait for adding an index. */
+        WAIT_ADD_INDEX("wait for adding an index to a record meta-data"),
```
```suggestion
        WAIT_ADD_INDEX("wait for adding an index"),
```

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
+        /** Wait for adding an index. */
+        WAIT_ADD_INDEX("wait for adding an index to a record meta-data"),
+        /** Wait for dropping an index. */
+        WAIT_DROP_INDEX("wait for dropping an index from a record meta-data"),
```
```suggestion
        WAIT_DROP_INDEX("wait for dropping an index"),
```

alecgrieser commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java:
```diff
@@ -248,6 +248,12 @@ public String title() {
+        /** Wait for dropping an index. */
+        WAIT_DROP_INDEX("wait for dropping an index from a record meta-data"),
+        /** Wait for updating records descriptor. */
+        WAIT_UPDATE_RECORDS_DESCRIPTOR("wait for updating the records descriptor in a record meta-data"),
```
```suggestion
        WAIT_UPDATE_RECORDS_DESCRIPTOR("wait for updating the records descriptor"),
```

amouehsan commented:
Rebased against master, and merged. Please let me know if you have any further comments. Thanks.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
         context.asyncToSync(FDBStoreTimer.Waits.WAIT_SAVE_META_DATA, saveAndSetCurrent(metaDataProto));
     }
 
+    private CompletableFuture<RecordMetaDataProto.MetaData> loadCurrentProto() {
```
I filed an issue for this (#409), but now that I look into it closer, I'm not sure if changing this would buy us much? Both `loadAndSetCurrent` and `saveAndSetCurrent` require the serialized proto that they load. Unless you feel strongly about it, how about I close that issue?

alecgrieser replied:
Nah, I suppose it's fine. I might leave a comment in #409 when you close it about why it can't/shouldn't be done.

amouehsan commented on fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java:
```diff
@@ -480,6 +494,205 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
+            recordMetaDataBuilder.updateRecords(recordsDescriptor);
+
+            // Validate the evolution of the new record meta-data
+            evolutionValidator.validate(RecordMetaData.build(metaDataProto).getRecordMetaData(), recordMetaDataBuilder.getRecordMetaData());
```
ah, that's right. It happens in `saveAndSetCurrent`. Dropped it from here. In `RecordMetaDataBuilder::updateRecords` we do not have access to neither of old and new meta-data, unless we build them. Is that what you mean? I can make it optional, build them, but I'm not sure if that's the correct thing to do.

alecgrieser replied:
We can still call `evolutionValidator.validateUnion` on the two descriptors, right?

alecgrieser replied:
Or the two union descriptors rather

amouehsan replied:
we can do it for the union descriptors. Let me give it a shot.

amouehsan replied:
done.
